### ISSUE TITLE
questions about label generate

### ISSUE THREAD

**Issue by hitczr** [wvangansbeke/LaneDetection_End2End#24]
Excuse me, I have several questions as below:
    1. Did you generate the coefficient label on the 256x512 image or on the original 720x1280  image?
    2. Why did you  do the (1-y_prime) operation on y_prime when draw the lines? (in utils.py:  def draw_fitted_line)

**Comment by wvangansbeke** [comment]
Hi @hitczr,

1. The labels are generated for a normalized grid between [0 - 1]. So it should be independent from the image size.
2. I had to change the reference frame from normal coordinates to pixel coordinates. Since the y_prime values are normalized, you can do 1-y_prime to get the pixel coordinates. This is necessary when you want to plot the lines.

Kind regards,
Wouter

**Comment by hitczr** [comment]
Thanks you for your answer. I have anther two questions:
1. When do the perspective transform, how did you choose the eight points(such as (0.45, 0.3), (0.55, 0.3) in src and (0.45, 1) in dst)? Did the pionts have any physical meanings such as the width of the lane?
2. When you draw the lanes in the image, why did you use the transform matrix M_scaledup rather than M? And what does the operation on M mean when you calculate the M_scaledup?

Kind regards,
Max

**Comment by wvangansbeke** [comment]
Hi Max,

1. I indeed used more or less the egolane to determine the correspondence points. These points are normalized between 0 and 1. I know that the lane lines should be straight lines in the bird's eye view perspective. So the two destination points of the same lane line should have the same x value (=0.45 for the left lane line or 0.55 for the right lane line). 

2. The perspective transformation matrix is normalized, since I normalized the correspondence points between 0 and 1. To draw on the image, I had to scale-up the transformation matrix. The reason why I do the normalization in the first place, is to make training less sensitive to hyperparameter tuning when using a different image size as input for the network.

Kind regards,
Wouter

**Comment by hitczr** [comment]
Hi, Wouter,
Now I have another question about valid:
    when I do the train and valid, all is run on the correct way, but when I set evaluate=True to run valid only, the predicted lane line shown on the picture was wrong. What's the problem? Do I need to change some other parameters?
Kind regards,
Max

**Comment by wvangansbeke** [comment]
Hi Max,

It might be that the best or latest model was not saved when you were training. You might have to check this. You can easily verify the loss. Normally the hyperparameters should stay the same.

Best,
Wouter

**Comment by hitczr** [comment]
Hi, Wouter, 
I have a question about the ground truth of params. When you calculate the params of the line on the bird-view image on Tusimple Dataset, did you do the y=1-y opreation to get the params[a, b, c]?

Kind regards, 
Max

**Comment by wvangansbeke** [comment]
Hi Max,

The coefficients are not calculated in pixel coordinates. You will need to do y=1-y when you want to plot the points or lines, since here you need pixel coordinates.

Maybe try plotting the ground truth lane lines first. 

Best, 
Wouter

**Comment by hitczr** [comment]
Hi, Wouter,
thanks you for your answer. I understand the operation of 1-y when draw lines. I have another question of generate params labels:
It seems that the curve drawn by the coefficient is the origin of the coordinates of the lower left corner of the image. Is that so?
    If so, My understand is:
        The normalized coordinates are the origin of the image in the upper left corner of the image. So, when i want to generate the ground truth of the params, I need to do the 1-y operation on the normalized coordinates to adapt the coordinates to the coordinate system with the lower left corner as the coordinate origin.
     Or, do you have any code about how to generate the params' ground truth that can share? I'm confused about this.
 Kind regards, 
Max

**Comment by wvangansbeke** [comment]
Hi Max,

The normalized coordinates and homography matrix are indeed in pixel coordinates except the ground truth. The ground truth has the lower left point as the origin. Are you maybe missing [this line](https://github.com/wvangansbeke/LaneDetection_End2End/blob/master/Networks/LSQ_layer.py#L107)?
Here I do the transformation `y=1-y` to get to the same reference frame as the ground truth. 

To get the coefficients: 1. Get you x, y pixel coordinates, 2. Do the homogeneous transformation, 3. transform back and normalize, 4. Do the fitting.
I will maybe put some code online, since I get a lot of questions about the exact computations.

Best,
Wouter

**Comment by hitczr** [comment]
Hi, Wouter,
  Thanks you for your answer. Your idea about end-to-end lane detection is great. I have solved the problems. And I will do a conclusion about the problems that I met during the generate labels, training and test to share with everyone.

Kind regards, 
Max

**Comment by drematrix** [comment]
Hello, Max !
Can you please direct how to test the model on custom dataset ? I am an undergraduate student working on self-driving-car and would like to test this model for lane detection. 
Thanks and Kind Regards,
Divye

**Comment by hitczr** [comment]
> Hello, Max !
> Can you please direct how to test the model on custom dataset ? I am an undergraduate student working on self-driving-car and would like to test this model for lane detection.
> Thanks and Kind Regards,
> Divye

Hello, Divye,
    To do the test, you need a Class which I named 'TestDataset' to load your test data and a function 'test'. To write the code about the function 'test', you can follow the function 'validate' in main.py. 
    BTW, if you want to test on your own data, you'd better do the train on your data first.  I am do the train now and just test the data, maybe I will put my test code later.

Kind regards, 
Max

**Comment by hitczr** [comment]
Hi, Wouter, 
    I have do the train and test on my data, but I have a qusetion about the Net:
    What change do I need to do when I train different scenes of data(such as day and night or highway and city road) together? Because when train them together, the train loss will not drop, but when I train them seperately, the result looks good. 
    Do you do any experiments on this area?
Kind regards, 
Max

**Comment by wvangansbeke** [comment]
Hi Max,

I would also train them together. The capacity of the network should be enough to handle both day and night situations I presume. Can you maybe try to train first with the conventional cross-entropy loss or in combination with our loss. If this does not work then something else should be wrong.

Best, 
Wouter

**Comment by hitczr** [comment]
Hi,  Wouter,
The last question has been solved. The points which I chose to do the perspective transform were wrong. 
Now, I want to get the img of weights map like Figure 5 in your paper, what function can I use in your repo?

Kind regards,
Max

**Comment by wvangansbeke** [comment]
Hi Max,

Take a look [here] (https://github.com/wvangansbeke/LaneDetection_End2End/blob/master/Networks/utils.py#L92). This function visualizes the weight maps.

Best,
Wouter

**Comment by hitczr** [comment]
Hi, Wouter,

I know this function, but the img it shows is a little bit different from the weights map in your paper. Did you do any post process to get the img(like the result of segmentation) in Figure 5 in your paper?

Kind regards,
Max

**Comment by wvangansbeke** [comment]
Hi Max,

Oh I see. This figure was computed with a different architecture based on ResNet-blocks and no down-sampling, aiming to easily understand the concept the paper. The weight maps can look different with a different architecture. You can play with different architectures if you want, but keep in mind that it is also dataset dependent. Since the TuSimple dataset has a lot of metal dots to distinguish the different lanes, the visualizations will also look different. 

Regards,
Wouter

**Comment by hitczr** [comment]
Hi, Wouter,
    What I want is the image of coordinate weights maps in Figure5 your paper, not the weights maps of different lines. The image of coordinate weights maps in the train and valid result are a little bit different from the image in your paper.
How can I get the same image of coordinate weights in you paper? Does the image of coordinate weights map also computed with a different architecture based on ResNet-blocks and no down-sampling?

Kind regards,
Max

**Comment by wvangansbeke** [comment]
Hi Max,

The coordinate maps are not weight maps but fixed matrices. You can find them in the code. Figure 5 shows just the sum of the different weight maps (=output maps of the network) computed with the network described above.

Best,
Wouter

**Comment by hitczr** [comment]
I get it. Thanks!

Best,
Max




发自网易邮箱 iPhone 版
On 07/18/2019 15:39, Wouter Van Gansbeke wrote:

Hi Max,

The coordinate maps are not weight maps but fixed matrices. You can find them in the code. Figure 5 shows just the sum of the different weight maps (=output maps of the network) computed with the network described above.

Best,
Wouter

**Comment by yoga-0125** [comment]
> Hi Max,
> 
> The normalized coordinates and homography matrix are indeed in pixel coordinates except the ground truth. The ground truth has the lower left point as the origin. Are you maybe missing [this line](https://github.com/wvangansbeke/LaneDetection_End2End/blob/master/Networks/LSQ_layer.py#L107)?
> Here I do the transformation `y=1-y` to get to the same reference frame as the ground truth.
> 
> To get the coefficients: 1. Get your x, y pixel coordinates, 2. Do the homogeneous transformation, 3. transform back and normalize, 4. Do the fitting.
> I will maybe put some code online, since I get a lot of questions about the exact computations.
> 
> Best,
> Wouter

Hi Wouter,

Could you upload some code to show how to calculate the data in Curve_parameters.json? 
Thanks a lot.
